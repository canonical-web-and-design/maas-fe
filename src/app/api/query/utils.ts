import { useMemo } from "react";

import type { UseQueryResult } from "@tanstack/react-query";

type QueryHook<T> = () => UseQueryResult<T[], unknown>;

export const useItemsCount = <T>(useItems: QueryHook<T>) => {
  const { data } = useItems();
  return useMemo(() => data?.length ?? 0, [data]);
};
