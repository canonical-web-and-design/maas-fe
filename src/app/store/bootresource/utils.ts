import type { BootResource } from "./types";

export const splitResourceName = (
  name: BootResource["name"]
): { os: string; release: string } => {
  const split = name.split("/");
  return split.length === 2
    ? { os: split[0], release: split[1] }
    : { os: "", release: "" };
};

export const splitImageName = (
  name: string
): { os: string; arch: string; subArch: string; release: string } => {
  const split = name.split("/");
  return split.length === 4
    ? { os: split[0], arch: split[1], subArch: split[2], release: split[3] }
    : { os: "", arch: "", subArch: "", release: "" };
};
