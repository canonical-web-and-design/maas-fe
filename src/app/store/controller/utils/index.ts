export {
  isControllerDetails,
  isRack,
  isRegionAndRack,
  isRegion,
} from "./common";
export { FilterControllers, getControllerValue } from "./search";
