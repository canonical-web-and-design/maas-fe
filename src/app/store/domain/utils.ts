import type { Domain, DomainDetails } from "./types";
import { RecordType } from "./types";

/**
 * Whether a record type is an address record.
 * @param domain - The record type to check.
 * @returns Whether the record type is an address record.
 */
export const isAddressRecord = (recordType: RecordType): boolean =>
  [RecordType.A, RecordType.AAAA].includes(recordType);

/**
 * Whether a domain has a DomainDetails type.
 * @param domain - The domain to check
 * @returns Whether the domain is DomainDetails.
 */
export const isDomainDetails = (
  domain?: Domain | null
  // Use "rrsets" as the canary as it only exists for DomainDetails.
): domain is DomainDetails => !!domain && "rrsets" in domain;
