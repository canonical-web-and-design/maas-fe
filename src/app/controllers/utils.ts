import { ControllerHeaderViews } from "./constants";
import type { ControllerHeaderContent } from "./types";

import { getNodeActionTitle } from "app/store/utils";

/**
 * Get title depending on header content.
 * @param defaultTitle - Title to show if no header content open.
 * @param headerContentName - The name of the header content to check.
 * @returns Header title string.
 */
export const getHeaderTitle = (
  defaultTitle: string,
  headerContent: ControllerHeaderContent | null
): string => {
  if (headerContent) {
    const [, name] = headerContent.view;
    switch (name) {
      case ControllerHeaderViews.ADD_CONTROLLER[1]:
        return "Add controller";
      default:
        return getNodeActionTitle(name);
    }
  }
  return defaultTitle;
};
