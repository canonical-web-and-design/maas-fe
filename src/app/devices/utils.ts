import { DeviceHeaderViews } from "./constants";
import type { DeviceHeaderContent } from "./types";

import { getNodeActionTitle } from "app/store/utils";

/**
 * Get title depending on header content.
 * @param defaultTitle - Title to show if no header content open.
 * @param headerContentName - The name of the header content to check.
 * @returns Header title string.
 */
export const getHeaderTitle = (
  defaultTitle: string,
  headerContent: DeviceHeaderContent | null
): string => {
  if (headerContent) {
    const [, name] = headerContent.view;
    switch (name) {
      case DeviceHeaderViews.ADD_DEVICE[1]:
        return "Add device";
      default:
        return getNodeActionTitle(name);
    }
  }
  return defaultTitle;
};
