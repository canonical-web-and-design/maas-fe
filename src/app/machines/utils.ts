import { MachineHeaderViews } from "./constants";
import type { MachineHeaderContent } from "./types";

import { getNodeActionTitle } from "app/store/utils";

/**
 * Get title depending on header content.
 * @param defaultTitle - Title to show if no header content open.
 * @param headerContentName - The name of the header content to check.
 * @returns Header title string.
 */
export const getHeaderTitle = (
  defaultTitle: string,
  headerContent: MachineHeaderContent | null
): string => {
  if (headerContent) {
    const [, name] = headerContent.view;
    switch (name) {
      case MachineHeaderViews.ADD_CHASSIS[1]:
        return "Add chassis";
      case MachineHeaderViews.ADD_MACHINE[1]:
        return "Add machine";
      default:
        return getNodeActionTitle(name);
    }
  }
  return defaultTitle;
};
