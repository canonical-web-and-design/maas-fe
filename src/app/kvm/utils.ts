import { KVMHeaderViews } from "./constants";
import type {
  KVMSidePanelContent,
  KVMStoragePoolResource,
  KVMStoragePoolResources,
} from "./types";

import type { MachineSidePanelContent } from "app/machines/types";
import { getHeaderTitle as getMachineHeaderTitle } from "app/machines/utils";
import type { Pod } from "app/store/pod/types";
import { formatBytes } from "app/utils";

/**
 * Returns a string with the formatted byte value and unit, e.g 1024 => "1KiB"
 *
 * @param memory - the memory in bytes
 * @returns formatted memory string with value and unit
 */
export const memoryWithUnit = (memory: number): string => {
  const formatted = formatBytes(memory, "B", { binary: true });
  return `${formatted.value}${formatted.unit}`;
};

/**
 * Get header title depending on header content.
 * @param sidePanelContent - The currently selected header content.
 * @returns Header title.
 */
export const getFormTitle = (sidePanelContent: KVMSidePanelContent): string => {
  switch (sidePanelContent.view) {
    case KVMHeaderViews.ADD_LXD_HOST:
      return "Add LXD host";
    case KVMHeaderViews.ADD_VIRSH_HOST:
      return "Add Virsh host";
    case KVMHeaderViews.COMPOSE_VM:
      return "Compose";
    case KVMHeaderViews.DELETE_KVM:
      return "Delete";
    case KVMHeaderViews.REFRESH_KVM:
      return "Refresh";
    default:
      // We need to explicitly cast sidePanelContent here - TypeScript doesn't
      // seem to be able to infer remaining object tuple values as with string
      // values.
      // https://github.com/canonical/maas-ui/issues/3040
      const machineSidePanelContent =
        sidePanelContent as MachineSidePanelContent;
      return getMachineHeaderTitle("", machineSidePanelContent);
  }
};

/**
 * Calculate the amount of free storage in a storage pool.
 * @param resource - The storage pool resource to calculate free storage.
 * @returns Free storage in pool.
 */
export const calcFreePoolStorage = (resource: KVMStoragePoolResource): number =>
  resource.total - resource.allocated_other - resource.allocated_tracked;

/**
 * Convert a pod or cluster's storage pool resources object into a sorted array.
 * @param pools - The pod or cluster's storage pool resources object.
 * @param defaultPoolId - the default pool id of the pod.
 * @returns a sorted list of storage pools in the pod or cluster.
 */
export const getSortedPoolsArray = (
  pools: KVMStoragePoolResources,
  defaultPoolId?: Pod["default_storage_pool"]
): [name: string, resource: KVMStoragePoolResource][] => {
  const poolsArray = Object.entries<KVMStoragePoolResource>(pools);

  return poolsArray.sort(([nameA, dataA], [nameB, dataB]) => {
    if (defaultPoolId && "id" in dataA && "id" in dataB) {
      // Pools in pods will have an id. For this case we sort by default first
      // (as defined in pod.default_storage_pool) then by id.
      if (
        dataA.id === defaultPoolId ||
        (dataB.id !== defaultPoolId && dataA.id < dataB.id)
      ) {
        return -1;
      } else if (dataB.id === defaultPoolId || dataA.id > dataB.id) {
        return 1;
      }
      return 0;
    }

    // Pools in clusters do not have a single id, as they can span multiple
    // pods. For this case we just sort by name;
    if (nameA < nameB) {
      return -1;
    } else if (nameA > nameB) {
      return 1;
    }
    return 0;
  });
};
