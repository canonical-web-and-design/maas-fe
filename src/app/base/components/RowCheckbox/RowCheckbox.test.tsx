import RowCheckbox from "./RowCheckbox";

import { render, screen } from "@/testing/utils";

describe("RowCheckbox", () => {
  it("can show a label", () => {
    render(
      <RowCheckbox
        handleRowCheckbox={vi.fn()}
        item={null}
        items={[]}
        label="Check row"
      />
    );
    expect(
      screen.getByRole("checkbox", { name: /Check row/i })
    ).toBeInTheDocument();
  });

  it("can check if it should be selected via a function", () => {
    render(
      <RowCheckbox
        checkSelected={() => true}
        handleRowCheckbox={vi.fn()}
        item={null}
        items={[]}
        label="Check row"
      />
    );
    expect(screen.getByRole("checkbox", { name: /Check row/i })).toBeChecked();
  });
});
