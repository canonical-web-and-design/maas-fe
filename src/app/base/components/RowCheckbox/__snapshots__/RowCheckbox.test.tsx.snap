// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RowCheckbox renders 1`] = `
<Input
  checked={false}
  className="keep-label-opacity"
  id="mock-redux-js-nanoid-1"
  labelClassName="u-no-margin--bottom u-no-padding--top"
  onChange={[Function]}
  type="checkbox"
  wrapperClassName="u-no-margin--bottom u-nudge--checkbox"
/>
`;
